---
title: "RDS Assignment Markdown"
author: "B198096"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Generating pretty heat maps from gene expression data \
## Setup

This pipeline is designed to generate a pretty heat map of a subset of 
Gene expression data from a data file of expression values, using associated 
Annotation files to label the data for useful interpretation. 

Note that before the pipeline begins processing the data I need to import 
The data into the working directory. I performed this in the terminal 
command line at the beginning of the project. The command for this is below

Retrieving the files: \
cp /shared_files/RDS_assignment_files/s2249132_files.zip . \
unzip s2249132_files.zip \



All code, both with and without comments, can be accessed here: \
https://github.com/B198096-2022/RDS_Assignment \
 
\
To copy the code to your working directory from a terminal enter: \
git clone https://github.com/B198096-2022/RDS_Assignment \



##Formatting The Data \

The pipeline reads in four files:

data_all.csv contains the actual data \
gene_annotation.csv contains the gene names and gene types for all genes analysed \
sample_annotation.csv contains information on the sample groupings \
genelist_7.txt is a personalized list of the genes that this pipeline will analyse \

The files are first read in as tables
Then the gene and sample annotations are added to the data table to generate
a large, annotated data frame, after which the list of personalized genes is used
to extract the desired rows from the data frame to create a smaller data frame 
which is used for the heat maps. Lastly, the data is log transformed. An error
check is done beforehand to ensure that there are no zeros to cause errors for 
log transformation as described below. 


##Checking and Filtering Data \

The pipeline contains three functions for checking the data values and correcting
any undesired values in the data 

zero_test(df, find) is a function that takes a data frame as its first argument
It reads through every data point in the data frame and generates a list of 
every position where there is a zero, printing out this list and returning 
the list as the output of the function 
It can also be used to find any specified number or value within the data frame, 
searching for the input argument 'find'. The default find is 0 but can be passed
as any integer or string. 


mimimum_test(df, min) is a similar function that takes a data frame as its 
first argument and reads through every data point in the data frame to check if 
it is equal to or less than the passed argument "min", which has a default of 1. 
It generates a list of every position in the data frame that meets this criteria
and both prints and returns this list. 

change_values <- function(df, find, change) takes a data frame as its first 
argument, then will find any data points that match  the argument 'find' and 
change them to the value of the argument 'change', which have a default of 0 and
1, respectively. It then returns the corrected data frame.
So this function corrects any zero values to ones by default to avoid 
log transformation errors. But it can be re-used for other error checking. 


The pipeline checks whether there are any zeroes in the data and only calls 
these functions if there are zeros. The data used for this pipeline did not 
possess any zeros, so the functions were not needed.  

\
##Annotating The Data \

The heat maps are annotated to label the gene types and sample treatment groups. 
This is accomplished by generating two annotation data frames. One is for the 
genes and the other for the samples. They are generated usingg the information 
from the annotation files, and are taken as an argument for the pheatmap function. 


```{r cars, include=FALSE}
data_all <- read.table(file= "data_all.csv",header=T,sep=",")

gene_annotation <- read.table(file= "gene_annotation.csv",header=T,sep=",")

sample_annotation <- read.table(file= "sample_annotation.csv",header=T,sep=",")

data_all_df <- as.data.frame(data_all[1:200, 2:13])

zero_test <- function(df, find=0) {
  all_data_matrix <- as.matrix(df)
  count <- 0
  zero_positions <- c()
  for (i in all_data_matrix){
    count = count + 1
    if (i == find){
      print("Zero value found at position:")
      print(count)
      zero_positions <- c(zero_positions, count)
    }
  }
  return(zero_positions)
}


mimimum_test <- function(df, min=1) {
  all_data_matrix <- as.matrix(df)
  count <- 0
  min_positions <- c()
  for (i in all_data_matrix){
    count = count + 1
    if (i < min){
      print("Filter value:")
      print(i)
      print("Found at position:")
      print(count)
      min_positions <- c(min_positions, count)
    }
  }
  return(min_positions)
}



change_values <- function(df, find=0, change=1) {
  all_data_matrix <- as.matrix(df)
  find_matrix <- all_data_matrix <= find
  all_data_matrix[find_matrix] <- change
  all_data_df <- as.data.frame(all_data_matrix)
  return(all_data_df)
}


zeros <- zero_test(data_all_df)
filter_response = 'n'

if (is.null(zeros) == FALSE){
  print("Zero values found in data")
  filter_response <- readline(prompt="Filter Data to change Zeros to Ones? (y/n): ")
}


if (filter_response == "y"|| filter_response == "yes") {
  data_all_df <- change_values(all_data_df)
}


data_all_log <- log(data_all_df)


#Annotate data and format for pheatmap 
gene_list <- read.delim("genelist_7.txt")

gene_vec <- unlist(gene_list)

uniq_gene_vec <- sort(unique(gene_vec))

selected_genes <- data_all_log[uniq_gene_vec,1:12]

gene_names <- as.data.frame(gene_annotation[uniq_gene_vec,3:4])

labeled_data <- cbind(gene_names,selected_genes)

labeled_data_matrix <- as.matrix(labeled_data[,3:14])

rownames(labeled_data_matrix) = gene_names[,2]


#Generate pheatmaps 
library(pheatmap)

genetype_annotate <- data.frame("Type" = labeled_data$Type)

rownames(genetype_annotate) = rownames(labeled_data_matrix)

colnames(genetype_annotate) = "Gene Type"

treatment_annotate <- data.frame(Treatment = sample_annotation$TreatmentGroup)

rownames(treatment_annotate) = colnames(labeled_data_matrix)

colnames(treatment_annotate) = 'Treatment Group'


```



\
\
##Generating the heat maps \

There are four heat maps generated by this pipeline. The first two heat maps show
the data clustered both by genes and by samples, whereas the third and fourth are 
only clustered by genes. 

The first and second heat map only differ by the addition of axis labels, which
results in a slight shrinking of the heat map that might make it harder for 
some users to read, so both maps are presented here. 

The same is true of maps 3 and 4, with map 4 simply being a labeled version of 
map 3. 

A row font size of 6 was selected because it was the largest font size possible 
that kept the gene names non-overlapping while allowing a box height that 
would fit onto the graphical display window for markdown 

\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\

Heat map 1: Clustering by genes and samples

```{r}
pheatmap(labeled_data_matrix, scale='row', fontsize_row = 6, 
         annotation_row=genetype_annotate, 
         annotation_col=treatment_annotate, 
         main = "Log(expression) of selected genes", 
         treeheight_col = 10, treeheight_row = 30)
```
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
Heat map 2: Clustering by genes and samples (With axis labels)

```{r}
library(grid)
setHook("grid.newpage", function() pushViewport(viewport(x=1,y=1,width=0.9, 
                                                         height=1.0, name="vp", 
                                                         just=c("right","top"))), action="prepend")
pheatmap(labeled_data_matrix, scale='row', fontsize_row = 6, 
         annotation_row=genetype_annotate, 
         annotation_col=treatment_annotate, 
         main = "Log(expression) of selected genes", 
         treeheight_col = 5, treeheight_row = 30)
setHook("grid.newpage", NULL, "replace")
grid.text("Sample Name", x=0.32, y=0.1, gp=gpar(fontsize=16))
grid.text("Gene Name", x=-0.05, rot=90, gp=gpar(fontsize=16))
```
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
Heat map 3: Clustering by genes only

```{r}
pheatmap(labeled_data_matrix, scale='row', fontsize_row = 6, 
         annotation_row=genetype_annotate, 
         annotation_col=treatment_annotate, 
         main = "Log(expression) of selected genes", 
         cluster_cols = FALSE, treeheight_row = 30)
```
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
Heat map 4: Clustering by genes only (With axis labels)

```{r}
library(grid)
setHook("grid.newpage", function() pushViewport(viewport(x=1,y=1,width=0.9, 
                                                         height=1.0, name="vp", 
                                                         just=c("right","top"))), action="prepend")
pheatmap(labeled_data_matrix, scale='row', fontsize_row = 6, 
         annotation_row=genetype_annotate, 
         annotation_col=treatment_annotate, 
         main = "Log(expression) of selected genes", 
         cluster_cols = FALSE, treeheight_row = 30)
setHook("grid.newpage", NULL, "replace")
grid.text("Sample Name", x=0.32, y=0.1, gp=gpar(fontsize=16))
grid.text("Gene Name", x=-0.05, rot=90, gp=gpar(fontsize=16))

```


\
\
\
\
\
\
\
\
\



#Interpreting the heatmaps 

Heat maps are a data visualization tool that utilize a color scale to represent
values on a numerical scale. 

These heat maps are being used to show gene expression. As indicated in the legend, 
warmer colors correspond to higher levels of gene expression whereas colder colors 
indicate lower levels of gene expression. 

The data includes gene expression values from four different treatment groups. The 
experiment is therefore evaluating whether gene expression changes between groups 
as a result of their differential treatment. This change is observed as a difference 
in the expression level of a given gene across columns. More importantly, 
significantly affected genes will show substantially different expression levels 
between entire treatment groups. 

For example, the first eight genes in the heat maps show high expression in 
treatment group one, middling expression in groups two and three, and low 
expression in group four. This indicates that treatment one is likely increasing the 
expression of these genes whereas treatment four is likely decreasing the expression
of these genes, with treatments two and three having a mid affect or no affect on 
the expression of these genes. 

The clustering feature of the heat maps puts rows or columns that are more similar 
in their expression profile closer together. In the above example of the top 
eight genes it is clear that they get clustered together because they all show 
a similar pattern of high, mid, then low expression across groups 1, 2/3, and 4, 
respectively. 

The clustering of the samples also shows that treatment groups 2 and 3 have the 
most similar expression profiles across the set of genes, and that these two groups
collectively are more similar to group 1 than group 4, as group 4 has the most 
disparate expression profile from the rest of the groups. 



\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\
\

#The Raw Data 
Here is the data that was plotted in the heat maps. 
Note: It is log transformed. 

I'm aware that this data puts the report above the page suggestion, so think of 
this as the supplemental data. Only here to see if you want to look 
at it. 

I also chose to include an extra pair of graphs for view preference and 
place the graphs on individual pages to keep the pages from looking cluttered, 
so this artificially increased the page count 

```{r}
labeled_data
```

